STL_Extension: Add bisect_failures debugging utility

lrineau opened the pull request:
## Summary

This PR adds a new debugging utility `bisect_failures` to the STL_Extension package. This function helps identify minimal failing test cases from complex input data by using a bisection-like approach.

**Small feature: [bisect failures debugging utility](https://cgalwiki.geometryfactory.com/CGAL/Members/wiki/Features/Small_Features/bisect_failures_debugging_utility)**

## Features

- **Multiple failure type support**: Handles CGAL exceptions, std::exceptions, and exit codes
- **Automatic granularity adjustment**: Starts with ratio=0.5, automatically halves when no fault is found
- **Debug output**: Saves intermediate states (`current`, `bad`, `error`, `final_bad`)
- **Template-based design**: Works with any copyable/assignable data type

## Testing

Added comprehensive test suite (`test_bisect_failures.cpp`) with 14 tests:
- ✅ Edge cases (empty data, no failures)
- ✅ Single element failures (CGAL exception, std::exception, exit code)
- ✅ Pair combinations (all 2-element permutations)
- ✅ Triple combinations (all 3 failure types together)

All tests pass in both Debug and Release configurations.

## Use Case

This utility is particularly useful for:
- Debugging complex geometric algorithms with large input datasets
- Finding minimal reproducible test cases from failing inputs
- Isolating problematic elements in mesh processing, triangulation, etc.

## Example Usage

```cpp
auto get_size = [](const Mesh& m) { return m.number_of_faces(); };
auto simplify = [](Mesh& m, std::size_t start, std::size_t end) -> bool {
  // Remove faces from start to end
  for(std::size_t i = start; i < end && i < m.number_of_faces(); ++i) {
    m.remove_face(start);
  }
  return true;
};
auto run = [](const Mesh& m) -> int { return algorithm(m); };
auto save = [](const Mesh& m, const std::string& prefix) {
  std::ofstream out(prefix + ".off");
  out << m;
};

int result = bisect_failures(mesh, get_size, simplify, run, save);
```

## Files Changed

- `STL_Extension/include/CGAL/bisect_failures.h` (new) - Main utility header
- `STL_Extension/test/STL_Extension/test_bisect_failures.cpp` (new) - Test suite  
- `STL_Extension/test/STL_Extension/CMakeLists.txt` (modified) - Added test

## Checklist

- [x] Code compiles without errors (Debug and Release)
- [x] All tests pass
- [x] Comprehensive documentation with Doxygen comments
- [x] Example usage in documentation
- [x] Follows CGAL coding style conventions
- [x] Deal with false positive in the testsuite

## Release Management

* Affected package(s): `STL_Extension`
* Feature/Small Feature (if any): [bisect failures debugging utility](https://cgalwiki.geometryfactory.com/CGAL/Members/wiki/Features/Small_Features/bisect_failures_debugging_utility)
* Link to compiled documentation (obligatory for small feature) [doc](https://cgal.github.io/9118/v2/STL_Extension/group__PkgSTLExtensionUtilities.html#gad56d708faac9063d3de057394a6d8c52)
* License and copyright ownership: no change

lrineau committed:
STL_Extension: Add bisect_failures debugging utility

Add new bisect_failures template function to help identify minimal failing
test cases from complex input data. The utility uses a bisection-like approach
to iteratively simplify data and locate the smallest subset that causes failures.

Features:
- Handles multiple failure types: CGAL exceptions, std::exceptions, exit codes
- Automatic ratio adjustment (starts at 0.5, halves when no fault found)
- Saves intermediate states for debugging (current, bad, error, final_bad)
- Comprehensive test suite with 14 tests covering all combinations:
  * Edge cases (empty data, no failures)
  * Single element failures (each failure type independently)
  * Pair combinations (all 2-element permutations)
  * Triple combinations (all 3 failure types together)
  * Large datasets (up to 200 elements)

The test suite demonstrates complete combinatorial coverage of three
independent failure conditions and validates the algorithm's ability to
isolate minimal failing cases.

All tests pass in both Debug and Release configurations.

lrineau committed:
fix license: LGPL

lrineau committed:
bisect_failures: fix documentation

lrineau commented:
/force-build:v1

lrineau committed:
fix errors in the testsuite

afabri commented on STL_Extension/include/CGAL/bisect_failures.h:
```diff
@@ -0,0 +1,239 @@
+ *
+ * \param data The input data to bisect
+ * \param get_size_fn Function that returns the "size" of the data (e.g., number of elements).
+ * \param simplify_fn Function that simplifies the data by removing elements from [start, end).
```
```suggestion
 * \param simplify_fn Function that simplifies the data by removing elements from `[start, end)`.
```

lrineau commented on STL_Extension/include/CGAL/bisect_failures.h:
```diff
@@ -0,0 +1,239 @@
+//
+// Author(s)     : Laurent Rineau
+//
+// The documentation has been partially generated using Github Copilot.
```
Only the documentation. The code is mine.

lrineau committed:
address Andreas' review, except for STL_Extension vs Profiling_tools

afabri commented on STL_Extension/test/STL_Extension/test_bisect_failures.cpp:
```diff
@@ -0,0 +1,243 @@
+// Author(s)     : Laurent Rineau
+
+// Test partially written by Github Copilot
+
```
```suggestion
```

afabri commented on STL_Extension/doc/STL_Extension/examples.txt:
```diff
@@ -1,4 +1,5 @@
 /*!
+\example STL_Extension/bisect_failures_example.cpp
```
Please drop the suffix  `_example`. There are many without it in this directory.

lrineau committed:
Update STL_Extension/test/STL_Extension/test_bisect_failures.cpp

Co-authored-by: Andreas Fabri <andreas.fabri@geometryfactory.com>

lrineau committed:
fixes after Andreas's second review

afabri commented:
Fails in 50% of today's testsuites  https://cgal.geometryfactory.com/CGAL/testsuite/summary-6.2-Ic-40.html?package=STL_Extension_Examples

lrineau committed:
avoid failures in the testsuite

lrineau committed:
undef NDEBUG and CGAL_NDEBUG

lrineau reviewed (commented):
> Fails in 50% of today's testsuites  https://cgal.geometryfactory.com/CGAL/testsuite/summary-6.2-Ic-40.html?package=STL_Extension_Examples 

I have enforced that test to be used without NDEBUG and CGAL_NDEBUG.

afabri commented:
Segfault in  [CGAL-6.2-Ic-41](https://cgal.geometryfactory.com/CGAL/testsuite/CGAL-6.2-Ic-41/STL_Extension_Examples/TestReport_lrineau_Debian-stable-Release.gz)

lrineau committed:
use std::endl to flush cout

lrineau committed:
fix compilation error

lrineau committed:
explain why NDEBUG must not be defined

sloriot commented on STL_Extension/include/CGAL/bisect_failures.h:
```diff
@@ -0,0 +1,245 @@
+ * \return Exit code: 0 (EXIT_SUCCESS) if no failures found, non-zero otherwise
+ *
+ * The algorithm:
+ * 1. Tests the full data first to verify it fails and capture the failure pattern
```
what if it does not fail? directly exit?

lrineau committed:
commit suggestions

lrineau committed:
tweak SKIP message

lrineau committed:
WIP: rewrite the documentation after review

lrineau commented on STL_Extension/include/CGAL/bisect_failures.h:
```diff
@@ -0,0 +1,245 @@
+ * \tparam GetSizeFn Function object type: `std::size_t GetSizeFn(const InputData& data)`
+ * \tparam SimplifyFn Function object type: `bool SimplifyFn(InputData& data, std::size_t start, std::size_t end)`
+ * \tparam RunFn Function object type: `int RunFn(const InputData& data)`
+ * \tparam SaveFn Function object type: `void SaveFn(const InputData& data, const std::string& filename_prefix)`
```
@afabri If `*_fn` is too short, what would you call them?

lrineau committed:
fix error

lrineau committed:
remove the `*Fn`/`*_fn` suffixes

lrineau committed:
rewrite the doc after review

lrineau committed:
tie clog to cerr in bisect_failures

That will ensure that the `clog` buffer is flushed before any output to `cerr`.

lrineau committed:
fix the Github action demo.yml

sloriot reviewed (commented):
I like this version better! Still some occurrences of minimal and smallest remaining that I'd prefer being replaced with reduced for example.

afabri reviewed (changes requested):
I made suggestions for replacing minimal @lrineau

lrineau committed:
apply suggestions during reviews

lrineau reviewed (commented):
I have applied all of your suggestions in the last commit.

albert-github commented on STL_Extension/include/CGAL/bisect_failures.h:
```diff
@@ -0,0 +1,288 @@
+ * \param run Function that tests the data. Should return 0 (`EXIT_SUCCESS`) on success, non-zero on failure.
+ *            May also throw exceptions to indicate failure.
+ * \param notify Function that is called with the data at different stages. It can be used to save
+ *                the data to a file or output. Its second parameter `event`, of type `Bisection_event`,
```
@afabri would defer from the official release documentation (still using 1.9.6 and is one of the overnight builds as well), would be a nice enhancement possibility for the "build the documentation)_

lrineau replied:
The CI uses the same Doxygen version as the one we use for the official releases. So far, that is Doxygen 1.9.6.

We should bump the Doxygen version for CGAL-6.2 to a more recent one.

sloriot committed:
avoid conversion warning

lrineau committed:
fix a warning

Note that is important to use a signed type for the loop variable.

The warning is:

> C:\CGAL_ROOT\CGAL-6.2-Ic-103\include\CGAL\bisect_failures.h(195,40): warning C4267: 'argument': conversion from 'size_t' to 'int', possible loss of data

See https://cgal.geometryfactory.com/CGAL/testsuite/CGAL-6.2-Ic-103/STL_Extension_Examples/TestReport_Christo_MSVC-2026-Community-Release.gz

lrineau committed:
bisect failure: do not compare cgal_exception->message()

The `message()` of a CGAL exception might have been generated at runtime.
It makes no sense to compare messages: the CGAL assertion is already identified
by the filename, and line number (plus the expression that was tested).
